File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <stack>

using namespace std;

vector<int> solution(vector<int> numbers) {
    vector<int> answer(numbers.size());
    stack<int> tmp;
    int curMax = 0;
    for(int i=numbers.size()-1;i>=0;i--){
        while(true){
            if(tmp.empty()){
                answer[i]=-1;
                break;
            }
            if(tmp.top()>numbers[i]){
                answer[i]=tmp.top();
                break;
            }
             tmp.pop();
        }
        tmp.push(numbers[i]);
        
    }
    return answer;
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Stack<Integer> st = new Stack<>();
        
        for(int i=0;i<numbers.length;i++){
            while(!st.empty()&&numbers[st.peek()]<numbers[i]){
                answer[st.peek()]=numbers[i];
                st.pop();
            }
            
                st.push(i);
        }
        while(!st.empty()){
            answer[st.peek()]=-1;
            st.pop();
        }

        return answer;
    }
}